tests/rados/test_osd_full.py

```diff
@@ -0,0 +1,322 @@
+    Autoscale PGs, ensure proper backfilling and rebalancing
+    1. Create replicated pool with single PG and autoscaling disabled
+    2. Retrieve the acting set for the pool
+    3. Calculate the amount of data to be written to pools in
+        order to trigger near-full, backfill-full, and completely
+        full OSD health warnings
+    4. Perform IOPS using rados bench to fill up the pool
+        up till nearfull, backfillfull and completely full capacity
+    5. Verify the number of objects written to the pool and the cluster
+        health warning after each IOPS iteration
+    6. Perform IOPS again and ensure IOPS was unsuccessful as OSDs are full
+    7. Increase the set-full-ratio to a higher value and set-backfillfull-ratio
+        to a value higher than the previous value of set-full-ratio
+    8. Ensure that completely full and backfillfull warning disappear and
+        nearfull warning remains
+    9. Enable autoscaling and observe the acting set OSD's 'used capacity' decreasing
+    10. Eventually nearfull warning should disappear, wait for PGs to achieve active+clean state
+    """
+
+    log.info(run.__doc__)
+    config = kw["config"]
+    cephadm = CephAdmin(cluster=ceph_cluster, **config)
+    rados_obj = RadosOrchestrator(node=cephadm)
+    pool_obj = PoolFunctions(node=cephadm)
+    client_node = ceph_cluster.get_nodes(role="client")[0]
+    bench_obj = RadosBench(mon_node=cephadm, clients=client_node)
+    osd_sizes = {}
+    bench_obj_size_kb = 16384
+    # object size for bench ops is chosen as 16 MB as this is the safe max value
+    # as per RHCS Administration Guide | benchmarking-ceph-performance_admin (https://red.ht/3zyuIrp)
+
+    log.info("Running pg rebalance with full OSDs test case")
+    pool_target_configs = config["pool_config"]
+
+    # Creating pools and starting the test
+    for entry in pool_target_configs.values():
+        try:
+            pool_name = entry["pool_name"]
+            log.info(
+                f"Creating {entry['pool_type']} pool on the cluster with name {pool_name}"
+            )
+            if entry.get("pool_type", "replicated") == "erasure":
+                method_should_succeed(
+                    rados_obj.create_erasure_pool, name=pool_name, **entry
+                )
+            else:
+                method_should_succeed(rados_obj.create_pool, **entry)
+
+            log.info(f"Pool {pool_name} created successfully")
+
+            # retrieving the size of each osd part of acting set for the pool
+            acting_set = rados_obj.get_pg_acting_set(pool_name=pool_name)
+            for osd_id in acting_set:
+                osd_df_stats = rados_obj.get_osd_df_stats(
+                    tree=False, filter_by="name", filter=f"osd.{osd_id}"
+                )
+                osd_sizes[osd_id] = osd_df_stats["nodes"][0]["kb"]
+                size_for_compare = osd_df_stats["nodes"][0]["kb"]
```

Comment:
The size_for_compare  value last OSD df stats not the acting_primary osd. It is better to compare with the primary osd.

Reply:
@SrinivasaBharath I have made changes to compare with the primary osd size, please review.

Reply:
LGTM